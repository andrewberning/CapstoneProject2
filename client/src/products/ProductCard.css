.ProductsCard {
  width: 250px; /* Set a fixed width */
  height: 400px; /* Set a fixed height */
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  margin: 10px; /* Add some margin for spacing */
}

.product-img-container {
  flex-grow: 1;
  display: flex;
  align-items: center;
  justify-content: center;
}

.card-img {
  max-width: 100%;
  max-height: 100%;
  object-fit: contain; /* Ensure the image fits within the container */
}

.card-body {
  flex-grow: 0;
  padding: 10px;
}

.btn {
  margin: 10px; /* Add margin for spacing */
}