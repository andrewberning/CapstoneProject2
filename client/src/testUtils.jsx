import React from "react";
import UserContext from "./auth/UserContext";
import { vi } from 'vitest';

const demoUser = {
  username: "testuser",
  first_name: "Test",
  last_name: "User",
  email: "testuser123@test.net",
};

// Mock functions
const mockSetCurrentUser = vi.fn();
const mockLogin = vi.fn();
const mockLogout = vi.fn();
const mockSignup = vi.fn();
const mockAddToCart = vi.fn();
const mockHasAddedToCart = vi.fn();

const UserProvider = ({ 
  children,
  currentUser = demoUser,
  setCurrentUser = mockSetCurrentUser,
  login = mockLogin,
  logout = mockLogout,
  signup = mockSignup,
  addToCart = mockAddToCart,
  hasAddedToCart = mockHasAddedToCart,
  cartItems = [],
  setCartItems = () => {},
  categories = [],
  infoLoaded = true,
  totalAmount = 0,
  totalItems = 0,
}) => (
  <UserContext.Provider value={{ 
    currentUser,
    setCurrentUser,
    login,
    logout,
    signup,
    addToCart,
    hasAddedToCart,
    cartItems,
    setCartItems,
    categories,
    infoLoaded,
    totalAmount,
    totalItems,
  }}>
    {children}
  </UserContext.Provider>
);

export { UserProvider, mockSetCurrentUser, mockLogin, mockLogout, mockSignup, mockAddToCart, mockHasAddedToCart };
