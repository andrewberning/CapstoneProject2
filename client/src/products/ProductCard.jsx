import { useContext, useEffect, useState } from "react";
import { Link } from "react-router-dom";
import UserContext from '../auth/UserContext';
import SignInModal from "../components/SignInModal";
import "./ProductCard.css"


export default function ProductCard({ id, name, price, image, category }) {
  const { currentUser, hasAddedToCart, addToCart, cartItems } = useContext(UserContext);
  const [added, setAdded] = useState(false);
  const [quantity, setQuantity] = useState(1);
  const [showModal, setShowModal] = useState(false);

  useEffect(() => {
    setAdded(hasAddedToCart(id));
  }, [id, hasAddedToCart, cartItems]);

  const handleQuantityChange = (evt) => {
    setQuantity(parseInt(evt.target.value, 10));
  };

  async function handleAddToCart(evt) {
    evt.preventDefault();
    if (!currentUser) {
      setShowModal(true);
      return;
    }
    if (added) return;
    const item = { id, name, price, image, category };
    await addToCart(currentUser, item, quantity);
    setAdded(true);
  }

  function handleCloseModal() {
    setShowModal(false);
  }
  
  return (
    <li className="ProductCard">
      <Link className="product-link card" to={`/${category}/${id}`}>
        <div className="product-img-container">
          <img src={image} className="card-img" />
        </div>
        <div className="card-body">
          <div className="product-item-details">
            <div className="card-title">{name}</div>
            <div className="card-price">${price}</div>
          </div>
        </div>
      </Link>

      {!added && (
        <div className="quantity-selector">
          <label htmlFor={`quantity-${id}`}>Quantity: </label>
          <select 
            id={`quantity-${id}`} 
            value={quantity} 
            onChange={handleQuantityChange}
          >
            {[...Array(10).keys()].map(i => (
              <option key={i + 1} value={i + 1}>
                {i + 1}
              </option>
            ))}
          </select>
        </div>
      )}
      
      <button 
        className="btn btn-primary mt-2"
        onClick={handleAddToCart} 
        disabled={added}
      >
        {added ? "Added to cart" : "Add to cart"}
      </button>
      <SignInModal show={showModal} handleClose={handleCloseModal} />
    </li>
  );
}