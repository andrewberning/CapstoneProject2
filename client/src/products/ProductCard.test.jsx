import { render } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import ProductCard from "./ProductCard";
import { UserProvider } from "../testUtils";

describe('ProductCard component', () => {
  it('renders without crashing', () => {
    render(
      <MemoryRouter>
        <UserProvider >
          <ProductCard />
        </UserProvider>
      </MemoryRouter>
    );
  });

  it("matches snapshot", function () {
    const { asFragment } = render(
      <MemoryRouter>
        <UserProvider>
          <ProductCard />
        </UserProvider>
      </MemoryRouter>,
    );
    expect(asFragment()).toMatchSnapshot();
  });
})