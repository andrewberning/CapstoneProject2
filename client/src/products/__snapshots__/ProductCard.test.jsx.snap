// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ProductCard component > matches snapshot 1`] = `
<DocumentFragment>
  <li
    class="ProductCard"
  >
    <a
      class="product-link card"
      href="/undefined/undefined"
    >
      <div
        class="product-img-container"
      >
        <img
          class="card-img"
        />
      </div>
      <div
        class="card-body"
      >
        <div
          class="product-item-details"
        >
          <div
            class="card-title"
          />
          <div
            class="card-price"
          >
            $
          </div>
        </div>
      </div>
    </a>
    <div
      class="quantity-selector"
    >
      <label
        for="quantity-undefined"
      >
        Quantity: 
      </label>
      <select
        id="quantity-undefined"
      >
        <option
          value="1"
        >
          1
        </option>
        <option
          value="2"
        >
          2
        </option>
        <option
          value="3"
        >
          3
        </option>
        <option
          value="4"
        >
          4
        </option>
        <option
          value="5"
        >
          5
        </option>
        <option
          value="6"
        >
          6
        </option>
        <option
          value="7"
        >
          7
        </option>
        <option
          value="8"
        >
          8
        </option>
        <option
          value="9"
        >
          9
        </option>
        <option
          value="10"
        >
          10
        </option>
      </select>
    </div>
    <button
      class="btn btn-primary mt-2"
    >
      Add to cart
    </button>
  </li>
</DocumentFragment>
`;
